Registering a component implementing multiple activiti listener interfaces throws exception on startup

<!--
Provide a general summary of the issue in the Title above

If you are reporting a new issue, make sure that we do not have any duplicates
already open. You can ensure this by searching the issue list for this
repository. If there is a duplicate, please close your issue and add a comment
to the existing issue instead.
-->
## I'm submitting a ...   (check one with "x")
```
[x] bug report
[ ] feature request
[ ] question
```

## Expected Behavior

I create a bean implementing `TaskListener` and `ExecutionListener`: https://gist.github.com/vierbergenlars/4b6aec9d4b1adcd629d9d730d2dd9906

This bean should be registered as both a task listener and as an execution listener in activiti.
<!--- If you're describing a bug, tell us what should happen -->
<!--- If you're suggesting a change/improvement, tell us how it should work -->
<!--- If you have a question, you can delete this -->

## Current Behavior

During startup of Alfresco, an exception is thrown, preventing Alfresco from starting.
```
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'typeBasedWorkflowRegistrar': Invocation of init method failed; nested exception is java.lang.IllegalStateException: overwrite of existing execution listener using id <XyzAuthenticationHandler>'
Caused by: java.lang.IllegalStateException: overwrite of existing execution listener using id <agfaMaterialsAuthenticationHandler>
	at com.github.dynamicextensionsalfresco.workflow.activiti.DefaultWorkflowTaskRegistry.registerTaskListener(DefaultWorkflowTaskRegistry.java:47)
	at com.github.dynamicextensionsalfresco.workflow.activiti.WorkflowTaskRegistrar.afterPropertiesSet(WorkflowTaskRegistrar.kt:49)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1573)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1511)
```

<!--- If describing a bug, tell us what happens instead of the expected behavior -->
<!--- If suggesting a change/improvement, explain the difference from current behavior -->
<!--- If you have a question, you can delete this -->

## Possible Solution

The cause of this exception is that the `AuthenticationHandler` class implements 2 interfaces.

When a bundle is started, beans that implement the interfaces `JavaDelegate`,  `TaskListener` and `ExecutionListener` are gathered one by one.

Then the instanciated classes will be registered to the relevant type depending on *all* interfaces they implement. https://github.com/xenit-eu/dynamic-extensions-for-alfresco/blob/master/annotations-runtime/shared/src/main/kotlin/com/github/dynamicextensionsalfresco/workflow/activiti/WorkflowTaskRegistrar.kt#L47-L51

This will result in the `AuthenticationHandler` being registered multiple times for the same types, which is not allowed by the `WorkflowTaskRegistry`


<!--- Not obligatory, but suggest a fix/reason for the bug, -->
<!--- or ideas how to implement the addition or change -->
<!--- If you have a question, you can delete this -->

## Steps to Reproduce (for bugs)
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->
<!--- If you have a question, you can delete this -->
1. Create a new DE bundle with https://gist.github.com/vierbergenlars/4b6aec9d4b1adcd629d9d730d2dd9906
2. Add the jar to an alfresco and start it up

## Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* Alfresco version used:
* DE version used:

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -29,6 +29,10 @@ Version template:
 ### Added
 * Alfresco 6 support
 
+### Fixed
+* [#189](https://github.com/xenit-eu/dynamic-extensions-for-alfresco/issues/189)
+Registering a component implementing multiple Activiti listener Interfaces throws exception on startup
+
 ## [1.7.6] - 2018-09-05
 * [#189](https://github.com/xenit-eu/dynamic-extensions-for-alfresco/issues/189) Support Json 2 on Alfresco 4
 

--- a/annotations-runtime/src/main/java/com/github/dynamicextensionsalfresco/workflow/activiti/WorkflowTaskRegistrar.java
+++ b/annotations-runtime/src/main/java/com/github/dynamicextensionsalfresco/workflow/activiti/WorkflowTaskRegistrar.java
@@ -1,6 +1,5 @@
 package com.github.dynamicextensionsalfresco.workflow.activiti;
 
-import java.beans.ExceptionListener;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
@@ -58,11 +57,11 @@ public class WorkflowTaskRegistrar implements InitializingBean, ApplicationConte
                 }
                 logger.debug("registered Bundle component {} (type: {}) -> {}", name, type, bean);
 
-                if (bean instanceof JavaDelegate) {
+                if (JavaDelegate.class.equals(type)) {
                     this.workflowTaskRegistry.registerDelegate(name, (JavaDelegate) bean);
-                } else if (bean instanceof TaskListener) {
+                } else if (TaskListener.class.equals(type)) {
                     this.workflowTaskRegistry.registerTaskListener(name, (TaskListener) bean);
-                } else if (bean instanceof ExecutionListener) {
+                } else if (ExecutionListener.class.equals(type)) {
                     this.workflowTaskRegistry.registerExecutionListener(name, (ExecutionListener) bean);
                 }
             });
@@ -77,11 +76,11 @@ public class WorkflowTaskRegistrar implements InitializingBean, ApplicationConte
                 activitiBeanRegistry.remove(name);
                 logger.debug("unregistered Bundle component {} (type: {}) -> {}", name, type, bean);
 
-                if (bean instanceof JavaDelegate) {
+                if (JavaDelegate.class.equals(type)) {
                     this.workflowTaskRegistry.unregisterDelegate(name);
-                } else if (bean instanceof TaskListener) {
+                } else if (TaskListener.class.equals(type)) {
                     this.workflowTaskRegistry.unregisterTaskListener(name);
-                } else if (bean instanceof ExecutionListener) {
+                } else if (ExecutionListener.class.equals(type)) {
                     this.workflowTaskRegistry.unregisterExecutionListener(name);
                 }
             });
@@ -107,6 +106,6 @@ public class WorkflowTaskRegistrar implements InitializingBean, ApplicationConte
     }
 
     private List<Class<?>> workflowDelegateTypes() {
-        return Arrays.asList(JavaDelegate.class, TaskListener.class, ExceptionListener.class);
+        return Arrays.asList(JavaDelegate.class, TaskListener.class, ExecutionListener.class);
     }
 }

--- a/annotations-runtime/src/test/java/com/github/dynamicextensionsalfresco/workflow/activiti/WorkflowTaskRegistrarTest.java
+++ b/annotations-runtime/src/test/java/com/github/dynamicextensionsalfresco/workflow/activiti/WorkflowTaskRegistrarTest.java
@@ -1,0 +1,78 @@
+package com.github.dynamicextensionsalfresco.workflow.activiti;
+
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.util.Collections;
+import java.util.HashMap;
+import org.activiti.engine.delegate.DelegateExecution;
+import org.activiti.engine.delegate.DelegateTask;
+import org.activiti.engine.delegate.ExecutionListener;
+import org.activiti.engine.delegate.TaskListener;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+import org.springframework.context.ApplicationContext;
+
+@RunWith(MockitoJUnitRunner.class)
+public class WorkflowTaskRegistrarTest {
+
+
+    @Mock
+    private WorkflowTaskRegistry workflowTaskRegistry;
+    @Mock
+    private ApplicationContext applicationContext;
+
+    private WorkflowTaskRegistrar workflowTaskRegistrar;
+
+    private static final String BEAN_ID_LISTENER = "TaskAndExceptionListener";
+    private TaskAndExceptionListener listener = new TaskAndExceptionListener();
+
+    @Before
+    public void setup() {
+        workflowTaskRegistrar = new WorkflowTaskRegistrar(new HashMap<>(), workflowTaskRegistry);
+        workflowTaskRegistrar.setApplicationContext(applicationContext);
+
+        when(applicationContext.getBeansOfType(TaskListener.class))
+                .thenReturn(Collections.singletonMap(BEAN_ID_LISTENER, listener));
+        when(applicationContext.getBeansOfType(ExecutionListener.class))
+                .thenReturn(Collections.singletonMap(BEAN_ID_LISTENER, listener));
+    }
+
+    @Test
+    public void afterPropertiesSet_beanImplementingMultipleActivitiListeners_correctlyRegistered() {
+        // https://github.com/xenit-eu/dynamic-extensions-for-alfresco/issues/199
+
+        workflowTaskRegistrar.afterPropertiesSet();
+
+        verify(workflowTaskRegistry).registerExecutionListener(BEAN_ID_LISTENER, listener);
+        verify(workflowTaskRegistry).registerTaskListener(BEAN_ID_LISTENER, listener);
+    }
+
+    @Test
+    public void destroy_beanImplementingMultipleActivitiListeners_correctlyRemoved() {
+        workflowTaskRegistrar.destroy();
+
+        verify(workflowTaskRegistry).unregisterExecutionListener(BEAN_ID_LISTENER);
+        verify(workflowTaskRegistry).unregisterTaskListener(BEAN_ID_LISTENER);
+    }
+
+    /**
+     * Class for testing purposes that implements multiple Activiti listener interfaces
+     */
+    private class TaskAndExceptionListener implements TaskListener, ExecutionListener {
+
+        @Override
+        public void notify(DelegateTask delegateTask) {
+
+        }
+
+        @Override
+        public void notify(DelegateExecution execution) {
+
+        }
+    }
+
+}

